lib/Sema/TypeCheckDecl.cpp

```diff
@@ -5581,6 +5583,152 @@ static Optional<ObjCReason> shouldMarkClassAsObjC(TypeChecker &TC,
   return None;
 }
 
+bool checkDeclCircularityImpl(Decl *decl, std::set<TypeBase *> &known);
+
+// Given a type, check if its in 'known'. Do the same with its subtypes.
+// When the type is nominal, get its declaration and check if any if its
+// member's type is in known by calling 'checkDeclCircularityImpl'
+bool isTypePresentForCircularity(Type type, std::set<TypeBase *> &known) {
+  auto desugared = type->getDesugaredType();
+
+  if (auto NT = dyn_cast<NominalType>(desugared)) {
+    if(auto NTT = NT->getDecl()->getDeclaredTypeInContext()) {
+      if (known.find(NTT.getPointer()) != known.end()) {
+        return true;
+      }
+    }
+    // the type is cleared, now we check its member.
+    return checkDeclCircularityImpl(NT->getDecl(), known);
+  }
+
+  // check each generic argument, this takes care of parameter from declaring
+  // context as well as stuff like Optionals
```

Comment:
I'm worried this is too restrictive. Right now, the following is valid in SILGen:

    struct R<T> { let t: () -> T }
    struct S { let r: R<S> }

I *think* with your code, that will be diagnosed as invalid.

Note that the condition here is that for a bound generic type `B<T1, T2, ...>`, it is recursive only if the generic parameter substituted for T1 is referenced from a storage member of B. The recursion may itself involve multiple levels of substitutions:

    struct RR<U> { let t: () -> T }
    struct G<U> { let u: U }
    struct R<T> { let r: G<RR<T>> }
    struct S { let r: R<S> }

The code in SIL type lowering applies substitutions when visiting types recursively, so for example, when we visit the member 'r' of 'S' in the last example above, we recursively walk members of R, applying the substitution `T -> S` to each member's type, which gives us `G<RR<S>>`.

Then we hit the bound generic type `G<RR<S>>`, recursively walk the members of G, with the substitution `U -> RR<S>`. We get to the member 't', which has substituted type () -> RR<S>. There's no more recursion to do (it's not a value type or tuple of value types), so we're OK -- we have not seen S, R, G, or RR appear recursively.

When you call getInterfaceType() on a member of a type, you get the type of that member without any substitutions applied.

Take a look at how the SIL code threads a substituted type through the recursion check, and when computing the type of a member, it uses an operation that takes the type decl, the substituted type, and the member, and applies the substitutions to the member's original type. This shouldn't be a big change to the logic here -- basically you're going to be passing either a TypeSubstitutionMap in addition to a Decl, or just the substituted Type of the Decl.

I hope this makes sense ;-) Please feel free to ask clarifying questions!